CONTRIBUTING.md

```diff
@@ -0,0 +1,151 @@
+[git](https://git-scm.com/), [Github](https://github.com/) and the dependency manager [composer](https://getcomposer.org/). 
+
+## First Step: Clone The Project And Install The Dependencies
+
+The phootwork project is a collection of single libraries that can be installed separately. Anyway, all the development
+process takes place in the global repository https://github.com/phootwork/phootwork .
+
+So, the first step is to fork the [phootwork repository](https://github.com/phootwork/phootwork) on Github and clone it
+on your local machine. If you have any doubt, please read https://help.github.com/en/articles/fork-a-repo .
+
+Then install the dependencies:
+```bash
+composer install
+```
+
+## Understanding The Directory Structure
+
+After cloning and installing the dependencies, you should have a directory structure like the following:
+```.
+   ├── bin
+   ├── src
+   │   ├── collection
+   │   ├── file
+   │   │   └── exception
+   │   ├── json
+   │   ├── lang
+   │   │   └── text
+   │   ├── tokenizer
+   │   └── xml
+   │       └── exception
+   ├── tests
+   │   ├── collection
+   │   │   └── fixtures
+   │   ├── file
+   │   ├── json
+   │   ├── lang
+   │   │   ├── fixtures
+   │   │   └── text
+   │   ├── tokenizer
+   │   │   └── fixtures
+   │   │       └── samples
+   │   └── xml
+   │       └── fixtures
+   └── vendor
+```
+
+Into the directory `bin` we have two scripts: `auto_split.sh` splits the whole Phootwork repository into the single
+library packages and `build_api.sh` generates the api documentation. You usually don't care about them, because they are
+run by our continuous integration server [Travis ci](https://travis-ci.org).
+
+As usual, into the directory `vendor` you can find all the libraries Phootwork depends on, both for runtime and development.
+
+The directory `src` contains the source code of our libraries and `tests` contains the relative tests.
+
+Both `src` and `tests` have sub-folders which correspond to the single libraries. E.g. you can find the source code of
+the library https://github.com/phootwork/lang into `src/lang` folder and the relative tests into `tests/lang` directory. 
+ 
+## Running The Test Suite
+
+While developing, the test part is very important: if you apply a patch to the existing code, the test suite must run without
+errors or failures and if you add a new functionality, no one will consider it without tests.
+
+Our test tool is [PhpUnit](https://phpunit.de/) and we provide a script to launch it:
+
+```bash
+composer test
+```
+Since our command runs phpunit binary under the hood, you can pass all phpunit options to it via the `--` operator, i.e.:
+```bash
+composer test -- --stop-on-failure
+```
+You can also use phpunit directly:
+```
+vendor\bin\phpunit
+```
+
+Each single library has its own test suite and you can launch it separately. I.e. if you want to run the `phootwork/lang`
+test suite:
+
+```bash
+composer test -- --testsuite lang
```

Comment:
I would have assumed, `composer test --testsuite lang` would be enough? Isn't it?

Reply:
Me too, but it isn't. If you write `composer test --testsuite lang` composer considers `--testsuite` a "composer option" and it raises an exception because it can't find this option.
On the contrary, everything after `--` is passed to the script.

Reply:
Oh wow, this is weird. Thanks for the explanation. Good to keep in mind.